Azure-Samples/azure-search-power-skills
# TextQualityWatchdog Function Definition, Model Training Notebook, and Dataset Creation Scripts

## Purpose

The changes include a new power skill, TextQualityWatchdog, that allows users to check the quality of text extracted during the document cracking process. Garbled/corrupt text will be detected by a pretrained machine learning model, and the corresponding document entry will be flagged in the index. Users can easily filter by the "text_quality_warning" field (0 for good quality, 1 for bad quality) to observe which documents could be problematic. This PR also includes utility scripts that can be used to build up a training dataset, as well as an ipynb demonstrating how to train, quantize, and evaluate the model. README.md is in progress.

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?


```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone https://github.com/WillChen789/azure-search-power-skills.git
```

* Test the code

```
After cloning the repo, deploy the TextQualityWatchdog project as an Azure function app (https://docs.microsoft.com/en-us/azure/azure-functions/create-first-function-vs-code-python) to be tested as part of a skillset in Cognitive Search. See the "sample.dat" file for an example of the expected input, if testing locally in VS Code. The demo notebook is meant to be run sequentially cell-by-cell to show how the model was trained and evaluated.
```

## What to Check
Verify that the following are valid
* TrainWatchdogModel.ipynb notebook should be able to run on CPU

## Other Information

The training labels were computed from Tika OOV scores for each document, which means that:
* A value of 0 for "text_quality_warning" should indicate that the extracted text is high quality
* A value of 1 for "text_quality_warning" usually indicates that the extracted text is low quality, but note that this is correlated with OOV text that is still well-formed
* Therefore some false positives are expected
